hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMCertStore.java

```diff
@@ -56,14 +60,14 @@
 /**
  * A Certificate Store class that persists certificates issued by SCM CA.
  */
-public class SCMCertStore implements CertificateStore {
+public final class SCMCertStore implements CertificateStore {
   private static final Logger LOG =
       LoggerFactory.getLogger(SCMCertStore.class);
   private SCMMetadataStore scmMetadataStore;
   private final Lock lock;
   private AtomicLong crlSequenceId;
 
-  public SCMCertStore(SCMMetadataStore dbStore, long sequenceId) {
+  private SCMCertStore(SCMMetadataStore dbStore, long sequenceId) {
```

Comment:
@bharatviswa504 After integrate with Ratis, the db writes should go to `transactionBuffer` and async flush to db by `takeSnapshot`, seems this part is missing.

Reply:
This was an intentional choice not to make CertStore DB updates use transactionalBuffer. The reason is, there is no in-memory map(Like how containerstateManager has), to get immediate reads work (Read operation after a write). (Or we should use table cache and add to table cache, and then listCerts need to be modified for it). As the request for certs in a cluster is a number of nodes in the cluster, so made this choice of directly persisting to DB.

When I am about to type it is idempotent, I see there is a problem, as for replay transaction because of below check looks 

    if ((getCertificateByID(serialID, VALID_CERTS) != null) ||
        (getCertificateByID(serialID, CertType.REVOKED_CERTS) != null)) {
      throw new SCMSecurityException("Conflicting certificate ID");
    }

If we move this check to DefaultCAServer, then storeValidCertificate will be an idempotent operation, and we will not have a replay issue. Let me know if this makes sense, I will open a Jira to fix this.

Reply:
Thanks for the verification!  Will inform you if find any issue.